modules/streaming/monix/src/main/scala/higherkindness/mu/rpc/healthcheck/handler/HealthServiceImpl.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2017-2019 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package higherkindness.mu.rpc.healthcheck.handler
+
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import higherkindness.mu.rpc.healthcheck._
+import higherkindness.mu.rpc.healthcheck.serviceMonix.HealthCheckServiceMonix
+import monix.execution.Scheduler
+import monix.reactive.{MulticastStrategy, Observable, Observer, Pipe}
+
+object HealthServiceMonix {
+
+  def buildInstance[F[_]: Sync](implicit s: Scheduler): F[HealthCheckServiceMonix[F]] = {
+
+    val checkRef: F[Ref[F, Map[String, ServerStatus]]] =
+      Ref.of[F, Map[String, ServerStatus]](Map.empty[String, ServerStatus])
+
+    val pipe: (Observer.Sync[HealthStatus], Observable[HealthStatus]) =
+      Pipe(
+        MulticastStrategy.behavior(
+          HealthStatus(new HealthCheck("FirstStatus"), ServerStatus("UNKNOWN"))))
+        .concurrent(s)
+
+    checkRef.map(c => new HealthCheckServiceMonixImpl[F](c, pipe))
+  }
+}
+
+class HealthCheckServiceMonixImpl[F[_]: Sync](
+    checkStatus: Ref[F, Map[String, ServerStatus]],
+    pipe: (Observer.Sync[HealthStatus], Observable[HealthStatus]))(implicit s: Scheduler)
+    extends AbstractHealthService[F](checkStatus)
+    with HealthCheckServiceMonix[F] {
+
+  override def setStatus(newStatus: HealthStatus): F[Unit] =
+    checkStatus
+      .update(_ + (newStatus.hc.nameService -> newStatus.status)) <*
+      Sync[F].delay(pipe._1.onNext(newStatus))
+
+  override def watch(service: HealthCheck): Observable[HealthStatus] =
+    pipe._2.filter(_.hc == service)
```

Comment:
I'm thinking about creating an implicit instance of `cats.kernel.Order` for the type `HealthCheck`, for example in a package object (`mu/modules/channel/src/main/scala/higherkindness/mu/rpc/healthcheck/package.scala`)

```scala
package higherkindness.mu.rpc

import cats.kernel.Order
import cats.instances.string._

package object healthcheck {

  implicit val orderForHealthCheck: Order[HealthCheck] = new HealthCheckOrder

  class HealthCheckOrder(implicit O: Order[String]) extends Order[HealthCheck] {

    override def eqv(x: HealthCheck, y: HealthCheck): Boolean =
      O.eqv(x.nameService, y.nameService)

    def compare(x: HealthCheck, y: HealthCheck): Int =
      O.compare(x.nameService, y.nameService)
  }

}
```

An then use the triple equals here:

```scala
pipe._2.filter(_.hc === service)
```

Reply:
What would this prevent? I'm not sure...

Reply:
Using the triple equals will check the types at compilation time, avoiding some bugs when this code will be evolved